### ISSUE TITLE
Should the tf.nn.sigmoid be replaced with tf.nn.sotfmax

### ISSUE THREAD

**Issue by mzhaoshuai** [ailias/Focal-Loss-implement-on-Tensorflow#9]
Thx for your code!
I am confused with some source code.

https://github.com/ailias/Focal-Loss-implement-on-Tensorflow/blob/master/focal_loss.py

`focal_loss.py`  line 22
```
sigmoid_p = tf.nn.sigmoid(prediction_tensor)
zeros = array_ops.zeros_like(sigmoid_p, dtype=sigmoid_p.dtype)
```
I think here we use `tf.nn.sigmoid` to calculate the probability is not suitable.

This is the sigmoid function:
![image](https://user-images.githubusercontent.com/22476764/43686260-009dc164-98f5-11e8-83ba-de246449d32c.png)

And this is the softmax function:
![image](https://user-images.githubusercontent.com/22476764/43686259-fbacc72c-98f4-11e8-9929-3272b84f0128.png)

Usually, we use softmax layer as the last layer to get the probability, so the sum is 1 and the margin between the output is increased.

**Comment by ChiefGodMan** [comment]
No, you'd better not. I have tested it and the loss cannot convergent. By the way, you can have a try.

**Comment by mzhaoshuai** [comment]
I tried it, the code is like this
```
sigmoid_pred = tf.nn.softmax(tf.clip_by_value(tf.to_float(logits), 1e-9, 1.0),  axis=-1)
sigmoid_pred = tf.clip_by_value(sigmoid_pred, clip_value_min=1e-9, clip_value_max=1.0)
num_classes = sigmoid_pred.shape[-1]
```
It works properly.

The reason that the loss do not converge is line 35
```
return tf.reduce_sum(per_entry_cross_ent)
```
![qq 20180808212148](https://user-images.githubusercontent.com/22476764/43839651-31e7b2ee-9b51-11e8-8c48-673fd6514196.png)


You do `reduce_sum`  instead of `reduce_mean` , this maybe not right. 
The picture is form 《Deep Learning》.
When we have a batch samples,  we may want to get the mean of the sum of their losses.

And why your original code is right? I think maybe the results of the sigmoid function is very small, then their sum is very small.So it works.

We do softmax because we want to get the probability, one propertity of the probability is that their sum is one. So I still think do softmax is right and do sigmoid is not reasonable.

**Comment by peter850706** [comment]
Hi @mzhaoshuai , I have the same thought with you, and I use softmax and "reduce_mean" but my result is very bad (my model always predict 0 for all input data after several iterations). Perhaps my data are severely imbalanced (foreground and background classes 1:100000 or worse), but I'm not sure the reason. Can you share your experience for setting its hyperparameter alpha, gamma or something?
By the way, I read the focal loss paper and I was wondering the following code:
```
per_entry_cross_ent = - alpha * (pos_p_sub ** gamma) * tf.log(tf.clip_by_value(sigmoid_p, 1e-8, 1.0)) \
                          - (1 - alpha) * (neg_p_sub ** gamma) * tf.log(tf.clip_by_value(1.0 - sigmoid_p, 1e-8, 1.0))
```
Should "pos_p_sub" and "neg_p_sub" be "1.0 - sigmoid_p" and "sigmoid_p"?

**Comment by mzhaoshuai** [comment]
Hi @peter850706 ,
I think the author other code is right.

line  28
```
pos_p_sub = array_ops.where(target_tensor > zeros, target_tensor - sigmoid_p, zeros)
```
line 32

```
neg_p_sub = array_ops.where(target_tensor > zeros, zeros, sigmoid_p)
```
The two lines do the same thing as `1.0 - sigmoid_p` and `sigmoid_p` respectively

I do not have much experience for setting its hyperparameter.
`gamma=2` is the original paper's recommoned number.

The main purpose of focal loss is to down-weight the easy classified samples.
And when alpha is larger, you may think this is not a easy positive sample, so you enlarge the alpha to ensure when this sample is classified wrongly, you will get a larger loss(relatively).
When you think this is not a easy negative sample, you enlarge the alpha.

But I think the simplest way is to try `alpha=0.25`,`alpha=0.5` and `alpha=0.75`.

And I think there maybe something wrong when your model always predict 0, setting parameter may do not help.

**Comment by peter850706** [comment]
@mzhaoshuai, thanks for your replies!

**Comment by ChiefGodMan** [comment]
Actually I had tested softmax and reduce_mean, but the result is not good because of the extreme small loss, so I have to change to sigmoid and reduce_sum, the paper "focal loss" explained the reason.
And, you can have a try for them.

**Comment by mzhaoshuai** [comment]
I think we may misunderstand the cross entropy loss.
1. https://pytorch.org/docs/master/nn.html?highlight=crossentropyloss#torch.nn.CrossEntropyLoss
![image](https://user-images.githubusercontent.com/22476764/45419889-2030aa80-b6bb-11e8-92ed-055311b38bca.png)

The cross entropy is come from the KL divergence.
1. https://rdipietro.github.io/friendly-intro-to-cross-entropy-loss/
2. https://datascience.stackexchange.com/questions/20296/cross-entropy-loss-explanation

For a multi classification problem. maybe the implement of the focal loss is like
1. https://github.com/clcarwin/focal_loss_pytorch/blob/master/focalloss.py

The original paper do say that sigmoid operation is more stable:
![image](https://user-images.githubusercontent.com/22476764/45420182-18bdd100-b6bc-11e8-9398-491164e9873e.png)

Recently I am busy.  Sorry for my late reply.
